[ROCm] Add bf16 starting from gfx11, bugfix & optimize RocmComputeCapability

Arech8 opened the pull request:
* Bugfix and improve device_description.h::RocmComputeCompatibility

* Enable ALG_DOT_BF16* on rocm with HW support

Arech8 committed:
Add bf16 starting from gfx11, bugfix & optimize RocmComputeCapability (#303)

* Bugfix and improve device_description.h::RocmComputeCompatibility

* Enable ALG_DOT_BF16* on rocm with HW support

(cherry picked from commit 510ea0628be3d3259def59f5a7b9a01811957474)

Arech8 commented:
@beckerhe JFYI: CI internal failure was fixed and it's fully green now.

beckerhe commented on xla/stream_executor/device_description.h:
```diff
@@ -54,13 +57,45 @@ class RocmComputeCapability {
+    return proto;
+  }
+
+  bool operator==(const RocmComputeCapability &other) const {
+    return gcn_arch_name_ == other.gcn_arch_name_;
+  }
+
```
Can you also define operator!= for symmetry reasons?

Arech8 replied:
Ok, I will, though it doesn't look to me that "symmetry reasons" is a valid reason to add a code for which there was no demand during all the lifetime of the project. This isn't a new class with an uncertain scope of use. It's a core component that was here from the inception of AMD support.

Arech8 replied:
added

beckerhe commented on xla/stream_executor/device_description.h:
```diff
@@ -54,13 +57,45 @@ class RocmComputeCapability {
   std::string gfx_version() const {
-    std::vector<std::string> tokens = absl::StrSplit(gcn_arch_name_, ':');
-    return tokens[0];
+    //  std::strchr() is faster for the case than std::string::find()
+    const char *const p_colon = std::strchr(gcn_arch_name_.c_str(), ':');
+    if (nullptr == p_colon) {
+      return gcn_arch_name_;  // likely it's the default invalid value
+    }
```
How is it relevant that this utility method is fast?

If this needs to be fast then we should do the split in the constructor and keep both parts around separately.

If you just want to make a bit faster we can avoid some temporary allocations:

```
std::pair<absl::string_view, absl::string_view> tokens = absl::StrSplit(gcn_arch_name_, ':');
return std::string(tokens.first);
```

But I'm a bit opposed to using raw C pointers and C library functions which makes the code harder to read.

beckerhe commented on xla/stream_executor/device_description.h:
```diff
@@ -69,98 +104,104 @@ class RocmComputeCapability {
 
-  bool operator==(const RocmComputeCapability &other) const {
-    return gcn_arch_name_ == other.gcn_arch_name_;
+  /// \brief Template-less implementation of IsThisGfxInAnyList().
+  /// \warning Don't use directly!
+  bool implIsThisGfxInAnyList(const absl::string_view *beg,
+                              const absl::string_view *end,
+                              const std::string &gfx) const {
+    return std::any_of(beg, end, [&gfx = gfx](const absl::string_view &s) {
+      return gfx == s;
+    });
```
Can you makes this a lambda expression in the actual function?

beckerhe commented on xla/stream_executor/device_description.h:
```diff
@@ -69,98 +104,104 @@ class RocmComputeCapability {
+  bool implIsThisGfxInAnyList(const absl::string_view *beg,
+                              const absl::string_view *end,
+                              const std::string &gfx) const {
+    return std::any_of(beg, end, [&gfx = gfx](const absl::string_view &s) {
```
Nit: I think it would be more idiomatic to just use `&` instead of `&gfx = gfx`.

beckerhe reviewed (commented):
Some minor comments, but in general it looks pretty straight forward to me.

beckerhe commented:
I just realized that there aren't any tests for any of the getters in RocmComputeCapability. Can you add those?

Arech8 committed:
add operator!= to RocmComputeCapability

I don't think it's useful, as there was never a demand for it, but a reviewer wants it.

Arech8 committed:
Add tests for main RocmComputeCapability accessors

Arech8 committed:
fix formatting of the test file

Arech8 committed:
Simplify gfx_version() default value test

Arech8 commented:
@beckerhe if you don't have capacity for the review, can you please assign someone else to do that?

Arech8 committed:
Restore gfx11_rx7900() and gfx12_rx8900() for AMD/TF

i-chaochen commented:
Hi @xla-rotation please review this PR?
